src/peft/optimizers/riemannian.py

```diff
@@ -0,0 +1,294 @@
+    model: PeftModel,
+    optimizer_cls: type[Optimizer],
+    optimizer_kwargs: dict,
+    lr_embedding: float = 1e-6,
+    reg: float = 1e-4,
+) -> Optimizer:
+    """
+    Creates a Riemmanian optimizer. Used only for LoRA. Implementation based on:
+    https://github.com/pilancilab/Riemannian_Preconditioned_LoRA. Paper reference: https://arxiv.org/pdf/2402.02347.
+
+    Args:
+        model (`torch.nn.Module`): The model to be optimized.
+        optimizer_cls (`torch.optim.Optimizer`): The optimizer class to be used.
+        optimizer_kwargs (`dict`): Additional keyword arguments to be passed to the optimizer.
+            - lr_embedding (`float`): The learning rate to be used for the embedding layer. Defaults to lr_embedding
+            - reg (`float`): Regularization parameter for Riemmanian preconditioner. Included for lora parameters only
+              and is needed for invertability guarantees
+    """
+
+    # TEST VERSION FOR ADAMW
+    if not issubclass(optimizer_cls, torch.optim.AdamW):
+        raise TypeError("TEST version only supports AdamW optimizer")
+    from ..tuners.lora.layer import Embedding
+
+    param_groups = {"lora_params": {}, "other_params": {}, "embedding": {}}
+
+    for name, param in model.named_parameters():
+        if not param.requires_grad:
+            continue
+        module = attrgetter(name)(model)
+        if isinstance(module, Embedding):
+            param_groups["embedding"][name] = param
+        elif "lora" in name:
+            param_groups["lora_params"][name] = param
+        else:
+            param_groups["other_params"][name] = param
+
+    lr = optimizer_kwargs["lr"]
+    weight_decay = optimizer_kwargs.get("weight_decay", 0.0)
+
+    optimizer_grouped_parameters = [
+        {
+            "params": list(param_groups["lora_params"].values()),
+            "weight_decay": weight_decay,
+            "lr": lr,
+            "is_lora": True,
+            "reg": reg,
+        },
+        {
+            "params": list(param_groups["embedding"].values()),
+            "weight_decay": weight_decay,
+            "lr": lr_embedding,
+            "is_lora": False,
+        },
+        {
+            "params": list(param_groups["other_params"].values()),
+            "weight_decay": weight_decay,
+            "lr": lr,
+            "is_lora": False,
+        },
+    ]
+
+    optimizer = RiemannianAdamW(optimizer_grouped_parameters, **optimizer_kwargs)
+    return optimizer
+
+
+class RiemannianAdamW(Optimizer):
+    """
+    This implements a variant of AdamW which combines with Riemannian preconditioner and is specifically designed for
+    LoRA model. Basic AdamW workflow follows implmentation of transformers.optimization.AdamW, with exception that
+    Riemmanian preconditioners are added. More specifically, for each LoRA parameter pair (lora_up, lora_down), we use
+    gradient of form
+                    grad(lora_up) = lora_up@inverse(lora_down.T@lora_down)
+    and vice versa.
+
```

Comment:
Let's mention that this is modified from the transformers AdamW implementation, not the torch one. It would also be nice to mark the lines in the function body that were changed compared to the original with a comment, if possible (probably everything inside of `if group["is_lora"]:`).